### ISSUE TITLE
I can't set guifonts

### ISSUE THREAD

**Issue by SR-MyStar** [neovide/neovide#1973]
### Neovide version:0.11.0.r0.gcb9bbe2-1(ArchLinuxCN) 0.10.4-2(extra)

```vim.o/opt.guifont="JetBrainsMono Nerd Font Mono,FZLanTingYuan-DB-GBK:h11"```
```vim.o/opt.guifont = { "JetBrainsMono Nerd Font Mono", "FZLanTingYuan-DB-GBK", ":h11" }```
And run "```neovide```" ```:set guifont=JetBrainsMono\ Nerd\ Font\ Mono,FZLanTingYuan-DB-GBK:h11```"
``
### They are all can't set guifonts
![2023-08-12-11-27-42](https://github.com/neovide/neovide/assets/139795416/4d75a9cb-20ea-4744-89ac-011835a28fb7)

# Log:

[neovide_r2023-08-12_15-33-05.restart-0000.log](https://github.com/neovide/neovide/files/12326855/neovide_r2023-08-12_15-33-05.restart-0000.log)

[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/12326856/neovide_rCURRENT.log)

## This is My Neovide configure
```
local M = {}

function M.init()
  M.config()
  if vim.g.transparent_background then
    M.background()
  end
  M.refresh(60, 5)
end

function M.config()
  -- 鼠标
  vim.o.mouse = "a"
  -- 行间距
  vim.opt.linespace = 2
  -- 字体
  vim.opt.guifont = { "JetBrainsMono Nerd Font Mono", "FZLanTingYuan-DB-GBK", ":h11" }
  -- 缩放
  vim.g.neovide_scale_factor = 0.9
  -- 填充
  vim.g.neovide_padding_top = 25
  vim.g.neovide_padding_bottom = 25
  vim.g.neovide_padding_right = 25
  vim.g.neovide_padding_left = 25
  -- 没有空闲
  vim.g.neovide_no_idle = true
  -- 退出需要确认
  vim.g.neovide_confirm_quit = true
  -- 是否全屏
  vim.g.neovide_fullscreen = false
  -- 记住以前窗口的大小
  vim.g.neovide_remember_window_size = true
  -- 使用super键位,比如<cmd>
  -- vim.g.neovide_input_use_logo = true
  -- 开启Alt和Meta按键
  vim.g.neovide_input_macos_alt_is_meta = true
  -- 触控板死亡地带
  -- vim.g.neovide_touch_deadzone = 6.0
  -- 触控板拖动超时
  -- vim.g.neovide_touch_drag_timeout = 0.17
  -- 开启光标粒子
  vim.g.neovide_cursor_vfx_mode = "railgun"
  -- vim.g.neovide_cursor_vfx_mode = "ripple"
  vim.g.neovide_cursor_vfx_particle_density = 20.0
  vim.g.neovide_cursor_animate_command_line = true
  vim.g.neovide_cursor_animate_in_insert_mode = true
end

function M.background()
  -- 设置透明背景
  local alpha = function()
    return string.format("%x", math.floor((255 * vim.g.transparency) or 0.8))
  end
  -- g:neovide_transparency should be 0 if you want to unify transparency of content and title bar.
  vim.g.neovide_transparency = 0.0
  vim.g.transparency = 0.0
  vim.g.neovide_background_color = "#0f1117" .. alpha()
end

-- 设置fps
function M.refresh(run_fps, free_fps)
  -- 设置刷新率
  vim.g.neovide_refresh_rate = run_fps
  -- 空闲刷新率
  vim.g.neovide_refresh_rate_idle = free_fps
end

return M
```
# [My neovim configure](https://github.com/SR-MyStar/MyNeovimDots)
## My Neovide configure file is in
```
MyNeovimDots
├──init.lua
└──lua
   └──neovide
      └──neovide.lua

```

**Comment by fredizzimo** [comment]
Try running `set guifont=*", that lists all the fonts available, and make sure that you copy paste the font name correctly from there.

**Comment by SR-MyStar** [comment]
I running ```set guifont=*``` ,My font(JetBrainsMono Nerd Font Mono) it's in this lists too,But it's not set guifont too.

**Comment by SR-MyStar** [comment]
I running set guifont=* ,My font(JetBrainsMono Nerd Font Mono) it's in this lists too,But I can't set guifonts too.

**Comment by SR-MyStar** [comment]
Font name is not spell error

**Comment by SR-MyStar** [comment]
**(it's not spell error)**

**Comment by fredizzimo** [comment]
Does running `:set guifont=JetBrainsMono\ Nerd\ Font\ Mono:h11` set it correctly? Without the fallback font? 

Can you also provide a shorter log with just staring the Neovide and then quit immediately. The important parts are not included in the logfile.

**Comment by SR-MyStar** [comment]
```
TRACE [neovide::redraw_scheduler] Next frame queued
TRACE [neovide::window::window_wrapper] Inner size: PhysicalSize { width: 937, height: 500 }
TRACE [neovide::redraw_scheduler] Next frame queued
DEBUG [neovide::renderer::fonts::caching_shaper] scale_factor changed: 0.90
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_size: 16.80px
TRACE [neovide::renderer::fonts::font_loader] Loading font FontKey { bold: false, italic: false, family_name: None, hinting: Full, edging: AntiAlias }
DEBUG [neovide::renderer::fonts::caching_shaper] Original font_width: 10.34px
DEBUG [neovide::renderer::fonts::caching_shaper] Font width: 10.34px (avg: 10.34px)
DEBUG [neovide::renderer::fonts::caching_shaper] Fudge factor: 0.97
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font size: 16.25px
DEBUG [neovide::renderer::fonts::caching_shaper] Fudged font width: 10.00px
TRACE [neovide::renderer::fonts::font_loader] Loading font FontKey { bold: false, italic: false, family_name: None, hinting: Full, edging: AntiAlias }
TRACE [neovide::renderer::grid_renderer] Updated font dimensions: Dimensions { width: 10, height: 23 }
```
[neovide_r2023-08-12_18-37-54.md](https://github.com/neovide/neovide/files/12327168/neovide_r2023-08-12_18-37-54.md)
[neovide_rCURRENT.md](https://github.com/neovide/neovide/files/12327169/neovide_rCURRENT.md)

**Comment by SR-MyStar** [comment]
Load fonts success,**but it's not my guifont option**

**Comment by fredizzimo** [comment]
I think we need to provide more logging, these are the important bits
```
DEBUG [neovide::renderer::fonts::caching_shaper] Updating font: JetBrainsMono Nerd Font Mono:h11
TRACE [neovide::renderer::fonts::font_loader] Loading font FontKey { bold: false, italic: false, family_name: Some("JetBrainsMono Nerd Font Mono"), hinting: Full, edging: AntiAlias }
ERROR [neovide::renderer::fonts::caching_shaper] Font can't be updated to: JetBrainsMono Nerd Font Mono:h11
```

Which does not tell the actual reason for the failure

**Comment by fredizzimo** [comment]
Also, could you try with `JetBrainsMono Nerd Font`, without the last mono part? I don't think you actually want that, because double width icons become too small. But it should still work of course.

**Comment by fredizzimo** [comment]
It looks like there's not much more logging we can do easily, but if you are able to build from source you could add 
`trace!("Typeface {:?}", typeface);` after this line
https://github.com/neovide/neovide/blob/cb9bbe2b0806c3febcf39b1cd9fbfde235c2b5ea/src/renderer/fonts/font_loader.rs#L79

But I'm betting that it fails before that, which means that Skia somehow fails to load the font. 

Does the font actually work in other applications?

**Comment by SR-MyStar** [comment]
This fonts it's working in other applications,My wezterm It's using this fonts too

**Comment by SR-MyStar** [comment]
I have new discoveries
```
❯ cargo run

    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `target/debug/neovide`
ERROR [neovide::renderer::fonts::caching_shaper] Font can't be updated to: JetBrainsMono Nerd Font Mono:h11
```

**Comment by SR-MyStar** [comment]
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/12327258/neovide_rCURRENT.log)

**Comment by fredizzimo** [comment]
Based on the logs, the problem seem is outside Neovide in the Skia library.

But I'm unable to repeat this myself, so we need someone that is able to repeat it and compiling from source, first try adding the `trace` as mentioned two posts above this one. And see if it prints something unexpected and if the line is reached at all.

That will much likely don't give any new information, and my educated guess is that this is a regression in the Skia version we just updated to. So, someone needs to try different versions, both newer beta versions and older to see if that's the case. That can be done by just editing https://github.com/neovide/neovide/blob/main/Cargo.toml#L84.

Alternatively, you could try to checkout Neovide 8944541f6dcbdbf52e5353db1a69a4298dcba630, which is before the upgrade.

**Comment by SR-MyStar** [comment]
That's so bad,All (Relaese)versions are can't using my font,I suspect it's the reason for the config file,Have you used my neovide config file?can or can't load fonts?

**Comment by SR-MyStar** [comment]
But,My configure files can set guifont in `neovim-qt`,So it isn't my configure error

**Comment by fredizzimo** [comment]
Sorry, I did not see that you already tried 0.10.4. Then it’s not a new issue caused by the updated skia. Can you try some other Nerd font, to check if that works? Did you already try`JetBrainsMono Nerd Font`, without the second `mono`?

**Comment by SR-MyStar** [comment]
Yes I try other fonts,They all can using in `neovim-qt` and my termianl but It's cat't using in `neovide` too

**Comment by fredizzimo** [comment]
I just tested with the official 0.11.0 Arch Linux package of Neovide, and for me `:set guifont=JetBrainsMono\ Nerd\ Font\ Mono` works as it should

I also tested the AUR version, and it works there as well.

**Comment by SR-MyStar** [comment]
I just added a new user for run :set guifont=JetBrainsMono\ Nerd\ Font\ Mono.but isn't works in the new user too.
That's so bad.This meant that I needed to spend a lot of time looking for the cause of this bug

**Comment by fredizzimo** [comment]
We understand that this can be frustrating, but there's no way for us to fix the issue without having someone with programming experience encounter the issue and debug it. 

If you can build from source and edit the code you can try to add the following line:
`trace!("Typeface {:?}", typeface);`

after this line
https://github.com/neovide/neovide/blob/cb9bbe2b0806c3febcf39b1cd9fbfde235c2b5ea/src/renderer/fonts/font_loader.rs#L79

And provide new logs. But note that even that is very unlikely to help us track down the bug. As far as I can see it happen's inside Google's Skia library https://github.com/google/skia.

They do have a new issue tracker here https://issues.skia.org/issues?q=is:open, but before we can report anything there, we need to at least be able to confirm that the bug happens inside Skia, with as many details as we can provide, at least I'm not going to report a bug there myself.

One thing you could try is another Skia based application like Google Chrome. Go to settings and customize font, and see if you can set it to use the Jetbrains mono nerd font there.

**Comment by SR-MyStar** [comment]
In Google Chrome,JetBrainMono Nerd Font works as it should, I don't know what to do track down the bug now,I'm tring build neovide from source and generated a new logs

**Comment by SR-MyStar** [comment]
This is new logfile
**cargo:v1.17.1
neovide(build):0.11.0**
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/12330023/neovide_rCURRENT.log)

**Comment by fredizzimo** [comment]
Thanks, can you add `trace!("Typeface {:?}", typeface);` on after line 79 in `neovide/src/renderer/fonts/font_loader.rs`

So that the function looks like this
```rust
    fn load(&mut self, font_key: FontKey) -> Option<FontPair> {
        let font_style = font_style(font_key.bold, font_key.italic);

        trace!("Loading font {:?}", font_key);
        if let Some(family_name) = &font_key.family_name {
            let typeface = self.font_mgr.match_family_style(family_name, font_style)?;
            trace!("Typeface {:?}", typeface);
            FontPair::new(font_key, Font::from_typeface(typeface, self.font_size))
        } else {
            let data = Data::new_copy(DEFAULT_FONT);
            let typeface = Typeface::from_data(data, 0).unwrap();
            FontPair::new(font_key, Font::from_typeface(typeface, self.font_size))
        }
    }
```
We need to know if `self.font_mgr.match_family_style()` fails or not, and if not what it returns. But after that it's harder, because I can see that we pass the correct things to Skia
```
TRACE [neovide::renderer::fonts::font_loader] Loading font FontKey { bold: false, italic: false, family_name: Some("JetBrainsMono Nerd Font"), hinting: Full, edging: AntiAlias }
ERROR [neovide::renderer::fonts::caching_shaper] Font can't be updated to: JetBrainsMono Nerd Font:b:h11
```

And I don't se any way to add more logging,  so Skia needs to be built from source and a debugger used to see why it's not loading the font.

**Comment by SR-MyStar** [comment]
> This is new logfile 这是新的日志文件 **cargo: 1.17.1 neovide(build): 0.11.0货物： 1.17.1 新维德（内部版本）： 0.11.0** [neovide_rCURRENT.log neovide_rCURRENT.log](https://github.com/neovide/neovide/files/12330023/neovide_rCURRENT.log)
 
I added `trace!("Typeface {:?}", typeface);` on after line 79 in `neovide/src/renderer/fonts/font_loader.rs` and sended neovide logfile

**Comment by SR-MyStar** [comment]
> This is new logfile
**cargo: 1.17.1
neovide(build): 0.11.0**
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/12330023/neovide_rCURRENT.log)

I added `trace!("Typeface {:?}", typeface);` on after line 79 in `neovide/src/renderer/fonts/font_loader.rs` and sended neovide logfile

**Comment by fredizzimo** [comment]
Thank you, it confirms that something goes wrong inside Skia. `matchFamilyName` returns null https://api.skia.org/classSkFontMgr.html#a91d8f2975af14c782a93d21236496272. And there's no error codes or anything to know what goes wrong. 

There's also no logging or anything in the internal call it should do here https://github.com/google/skia/blob/b8108412832504761a799be9a83f56a399d4f16c/src/ports/SkFontMgr_fontconfig.cpp#L829-L866 (on Linux it should use fontconfig). 

So the only way to debug this as far as I can see, is to build skia from source, attach a debugger and see what goes wrong on your machine, or some other machine that behaves the same way.

**Comment by SR-MyStar** [comment]
Emmm,I'm run some commands and then open neovide,and then ... neovide can set fonts now,That's so weird.
**run commands**
```
❯ mkfontdir
❯ mkfontscale
❯ ln -s ~/.local/share/fonts ./.fonts
```

**Comment by fredizzimo** [comment]
Ah, now I see. You have the fonts installed as user in `~/.local/share/fonts`, that seems to be the important thing missing from the issue. I will check later today if I can reproduce the issue that way. 

It didn't occur to me that you would have installed it as a user font, since there's an official package available https://archlinux.org/packages/extra/any/ttf-jetbrains-mono-nerd/

**Comment by SR-MyStar** [comment]
I'm installed `nerd-fonts-git` aur package,no install official's "[ttf-jetbrains-mono-nerd/](https://archlinux.org/packages/extra/any/ttf-jetbrains-mono-nerd)" package,user fonts it's copy from `/usr/share/fonts/nerd-fonts-git/JetBrain*.ttf`
